File: runtime-benchs/esp32-native/bm-heatdetect/main/heat_calib.c
#include "measure.h"
#include "heat_sensor_data.h"

int16_t *ACal;
int32_t *QCal;
int16_t *stdCal;
int16_t *zscore;
int16_t z_max, z_min;

void fast_calibration(uint16_t frame_buffer[64], uint16_t frame_number);
uint16_t isqrt (int32_t x);
void zscoreCalculation(uint16_t frame_buffer[64]);

// Split into separate function to avoid the compiler just optimising away the whole test.
void heat_calib() {
    // uint16_t frame_buffer[64];
    uint16_t* frame_buffer=malloc(64*sizeof(uint16_t));

    for (uint16_t i=0; i<100; i++) {
        get_heat_sensor_data(frame_buffer, i);
        fast_calibration(frame_buffer, i);
    }
    get_heat_sensor_data(frame_buffer, 100);
    zscoreCalculation(frame_buffer);
}

#define rounding_int_division(dividend, divisor) (((dividend) + ((divisor) / 2)) / (divisor))

void fast_calibration(uint16_t frame_buffer[64], uint16_t frame_number) {
    for(uint16_t i=0; i<64; i++) {
        uint16_t previous_ACal = ACal[i];
        ACal[i] += rounding_int_division((int32_t)frame_buffer[i] - ACal[i], frame_number+1);
        QCal[i] += ((int32_t)frame_buffer[i] - previous_ACal) * ((int32_t)frame_buffer[i] - ACal[i]);
    }
    for(uint16_t i=0; i<64; i++) {
        stdCal[i] = isqrt(QCal[i]/(frame_number+1));
    }
}

// http://www.cc.utah.edu/~nahaj/factoring/isqrt.c.html
uint16_t isqrt (int32_t x) {
    uint32_t   squaredbit, remainder, root;

    if (x<1) return 0;

    /* Load the binary constant 01 00 00 ... 00, where the number
    * of zero bits to the right of the single one bit
    * is even, and the one bit is as far left as is consistant
    * with that condition.)
    */
    squaredbit  = (uint32_t) ((((uint32_t) ~0L) >> 1) & 
                        ~(((uint32_t) ~0L) >> 2));
    /* This portable load replaces the loop that used to be 
    * here, and was donated by  legalize@xmission.com 
    */

    /* Form bits of the answer. */
    remainder = x;  root = 0;
    while (squaredbit > 0) {
        if (remainder >= (squaredbit | root)) {
            remainder -= (squaredbit | root);
            root >>= 1; root |= squaredbit;
        } else {
            root >>= 1;
        }
        squaredbit >>= 2; 
    }

    return root;
}

void zscoreCalculation(uint16_t frame_buffer[64]) {
    int32_t tempMax = -300000;
    int32_t tempMin = 300000;

    for(int i=0; i<64; i++) {
        int16_t score = 100 * ((int32_t)frame_buffer[i] - (int32_t)ACal[i]) / stdCal[i];

        zscore[i] = score;

        if(score > tempMax) {
            tempMax = score;
        }

        if(score < tempMin) {
            tempMin = score;
        }
    }

    z_max = tempMax;
    z_min = tempMin;
}

File: runtime-benchs/esp32-micropython/heatcalib.py
import time
global start_time
def rtc_startBench():
    global start_time
    start_time = time.ticks_ms()


def rtc_stopBench():
    global start_time
    delta = time.ticks_diff(time.ticks_ms(),start_time)
    print(delta)


import random
import math

ACal = [0] * 64
QCal = [0] * 64
stdCal = [0]*64
zscore = []
heat_sensor_data = []

def generate_sensor_data():
    for i in range(0, 101):
        l = []
        for j in range(0, 64):
            r = random.random()
            if r < (1/3):
                l.append(2575)
            elif r < 2/3:
                l.append(2600)
            else:
                l.append(2625)
        heat_sensor_data.append(l)
    # print(heat_sensor_data)


def get_heat_sensor_data(frame_buffer, frame_number):
    for i in range(0, 64):
        frame_buffer.append(heat_sensor_data[frame_number][i])
    return frame_number


def heat_calib():
    frame_buffer = []
    for i in range(0, 100):
        get_heat_sensor_data(frame_buffer, i)
        fast_calibration(frame_buffer, i)
    get_heat_sensor_data(frame_buffer, 100)
    zscoreCalculation(frame_buffer)


def rounding_int_division(dividend, divisor):
    return int((dividend + divisor / 2) / divisor)


def fast_calibration(frame_buffer, frame_number):
    global stdCal
    global ACal
    global QCal
    for i in range(0, 64):
        previous_ACal = ACal[i]
        ACal[i] += rounding_int_division(frame_buffer[i] - ACal[i],
                                         frame_number + 1)
        QCal[i] += (frame_buffer[i] - previous_ACal) * (
                    frame_buffer[i] - ACal[i])

    stdCal = [math.isqrt(int(QCal[i] / (frame_number + 1))) for i in range(0, 64)]
    for i in range(0, 64):
        if stdCal[i] == 0:
            stdCal[i] = 1

def zscoreCalculation(frame_buffer):

    tempMax = -300000
    tempMin = 300000

    for i in range(0, 64):
        score = 100 * (int(frame_buffer[i]) - int(ACal[i])) / stdCal[i]

        zscore.append(score)

        if (score > tempMax):
            tempMax = score

        if (score < tempMin):
            tempMin = score

    z_max = tempMax
    z_min = tempMin



generate_sensor_data()
rtc_startBench()

heat_calib()
rtc_stopBench()

